### PR TITLE
Add support for building inside a docker container.

### PR DESCRIPTION
**Pull request by STRML**
This simplifies the initial build, especially for users who want to clone & build a new binary off master or with a specific fix.

### CODE REVIEW

File: Dockerfile.build [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM golang:1.6
+
+ENV GO15VENDOREXPERIMENT=1
```
Comment (tboerger) [inline Dockerfile.build]: Drop that, it's the default on 1.6

File: Dockerfile.build [hunk 2]
```diff
@@ -0,0 +1,13 @@
+FROM golang:1.6
```
Comment (tboerger) [inline Dockerfile.build]: Maybe use the alpine image?

File: Dockerfile.build [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+ENV GO15VENDOREXPERIMENT=1
+ENV GOPATH=/go
```
Comment (tboerger) [inline Dockerfile.build]: This variable is already part of the base image.

File: Dockerfile.build [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+ENV PATH=$PATH:$GOPATH/bin
```
Comment (tboerger) [inline Dockerfile.build]: This should also be already set

File: Dockerfile.build [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+COPY . /go/src/github.com/drone/drone
+WORKDIR /go/src/github.com/drone/drone
```
Comment (tboerger) [inline Dockerfile.build]: I would move this one line up since it's more fixed than the copy just to reuse this layer.

File: Dockerfile.build [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+RUN make deps
```
Comment (tboerger) [inline Dockerfile.build]: Since when does the go base image include make?

Comment (tboerger) [inline Dockerfile.build]: The deps and gen task should also be part of the cmd.

File: README.md [hunk 7]
```diff
@@ -34,21 +34,12 @@ Please see our [installation guide](http://readme.drone.io/setup/overview) to in
...
-Commands to build from source:
+Build from source (artifacts will be in `./release`):
 
 ```sh
-export GO15VENDOREXPERIMENT=1
-
-make deps    # Download required dependencies
-make gen     # Generate code
-make build   # Build the binary
+git clone git://github.com/drone/drone.git
```
Comment (tboerger) [inline README.md]: Not everybody wants to build it in docker, so extend the readme but don't drop the current guide.

File: Dockerfile.build (hunk 6 above)
Comment (vaijab) [inline Dockerfile.build]: golang image does have `make` actually.

**Comment by STRML** [comment]
Thanks for the comments - will address.

File: Dockerfile.build (hunk 6 above)
Comment (STRML) [inline Dockerfile.build]: Point of separating them out was to generate intermediate images so you could re-run the build more easily (I'll have to separate the `ADD` statements to make this work properly, actually).

**Comment by STRML** [comment]
Committed based on Alpine, updated README, set `.dockerignore` move to move only if it doesn't exist (so multiple failed runs will still restore the correct `.dockerignore`).

File: Dockerfile.build (hunk 2 above)
Comment (appleboy) [inline Dockerfile.build]: Why not use the latest go version 1.7.3?

**Comment by bradrydzewski** [comment]
I'm wondering if a Dockerfile.build is necessary. I think one could just run the following command (or something like it) to build drone inside a container and have the artifacts output on the host machine

```
docker run --rm --volume=${pwd}:/go/src/github.com/drone/drone golang:1.7 make deps gen build
```

**Comment by vaijab** [comment]
> and have the artifacts output on the host machine

That's the whole point. I also think that docs update is sufficient.

**Comment by STRML** [comment]
Sharing a volume in that way won't work normally on osx/windows without
additional configuration. But the command could certainly be made shorter
if we're willing to stick with the regular go image over the alpine one.

**Comment by bradrydzewski** [comment]
If you are using docker for mac (as opposed to older boot2docker) it mounts the volume from the host into the container. This command works for me on osx, compiling to `release/` on the host machine:

```
docker run \
    -w /go/src/github.com/drone/drone \
    -v $(pwd):/go/src/github.com/drone/drone \
    golang:1.6 make deps gen build
```

It is also possible to build drone using the drone command line utility. This will mount the working directory as a volume, and run the drone pipeline steps, compiling binaries to `release/`

```
drone exec
```

I definitely appreciate what this pull request is trying to do. I also see from the comments that everyone has an opinion about how this should be done. This leads me to believe that drone should provide one canonical way to build the project (Makefile) and let the individual developer modify or augment the build process with the tools and technologies of their choosing.

For these reasons I would prefer not to provide alternate methods as part of the official drone build process. I think this thread will remain useful to individuals that are looking for instructions to compile in Docker, and can use it to make informed decisions on the approach that best suits their workflow.

**Pull request closed without merging**